lib/adapters/binance.py

```diff
@@ -0,0 +1,166 @@
+                    'l': "LowPrice",
+                    'v': "TotalTradedBaseAssetVolume",
+                    'q': "TotalTradedQuoteAssetVolume",
+                    'O': "StatisticsOpenTime",
+                    'C': "StatisticsCloseTime",
+                    'F': "FirstTradeId",
+                    'L': "LastTradeId",
+                    'n': "TotalNumberOfTrades",
+                }[col],
+                inplace=True)
+            df.set_index('EventTime', inplace=True)
+            df.index = pd.to_datetime(df.index, unit='ms')
+            return df
+
+
+    class BinanceUserPayloadTransformer:
+        """ """
+
+        def __call__(self, payload):
+            try:
+                message_map = {
+                    "outboundAccountInfo": self.account_update,
+                    "outboundAccountPosition": self.position_update,
+                    "balanceUpdate": self.balance_update,
+                    "executionReport": self.order_report
+                }
+
+                message_map[payload["e"]](payload)
+            except KeyError:
+                msg = "unknown payload type {p}".format(p=payload.get("e"))
+                raise InvalidPayloadRecieved(msg)
+            except Exception as e:
+                raise Exception("unknown error occured in user stream", e)
+
+
+        def initial_rest_payload(self, payload) -> AccountStatus:
+            return AccountStatus(
+                CanWithdraw=payload["canWithdraw"],
+                CanTrade=payload["canTrade"],
+                CanDeposit=payload["canDeposit"],
+                Positions={
+                    balance["a"]: Position(
+                        Locked=float(balance["l"]),
+                        Free=float(balance["f"])
+                    ) for balance in payload["B"]
+                }
+            )
+
+
+        def account_update(self, payload) -> Tuple[str, AccountStatus]:
+            return UpdateType.ACCOUNT, AccountStatus(
+                CanWithdraw=payload["W"],
+                CanTrade=payload["T"],
+                CanDeposit=payload["D"],
+                Positions={
+                    balance["a"]: Position(
+                        Locked=float(balance["l"]),
+                        Free=float(balance["f"])
+                    ) for balance in payload["B"]
+                }
+            )
+
+
+        def balance_update(self, payload) -> Tuple[str, BalanceUpdate]:
+            return UpdateType.BALANCE, BalanceUpdate(
+                Asset=payload["a"], Update=float(payload["d"])
+            )
+
+
+        def position_update(self, payload) -> Tuple[str, PositionStatus]:
+            return UpdateType.POSITION, {
+                balance["a"]: Position(
+                    Locked=float(balance["l"]),
+                    Free=float(balance["f"])
+                ) for balance in payload["B"]
+            }
+
+
+        def order_report(self, payload):
+            # @todo - never did work out how to handle these.
```

Comment:
This should be forwarded on to the `Account` regardless, as I envision logging of transactions/orders to be done via the `Account`.